src/components/tl/dpu/tl_dpu_context.c

```diff
@@ -0,0 +1,246 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "tl_dpu.h"
+#include "tl_dpu_coll.h"
+
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <netinet/ip.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/types.h>
+
+#include <netdb.h>
+#include <poll.h>
+#include <errno.h>
+#include <unistd.h>
+
+static void err_cb(void *arg, ucp_ep_h ep, ucs_status_t status)
+{
+    fprintf(stderr, "error handling callback was invoked with status %d (%s)\n",
+                    status, ucs_status_string(status));
+}
+
+static int _server_connect(char *hname, uint16_t port)
+{
+    int sock = 0, n;
+    struct addrinfo *res, *t;
+    struct addrinfo hints = { 0 };
+    char service[64];
+
+    hints.ai_family   = AF_UNSPEC;
+    hints.ai_socktype = SOCK_STREAM;
+    sprintf(service, "%d", port);
+    n = getaddrinfo(hname, service, &hints, &res);
+
+    if (n < 0) {
+        fprintf(stderr, "%s:%d: getaddrinfo(): %s for %s:%s\n", __FILE__,__LINE__, gai_strerror(n), hname, service);
+        return -1;
+    }
+
+    for (t = res; t; t = t->ai_next) {
+        sock = socket(t->ai_family, t->ai_socktype, t->ai_protocol);
+        if (sock >= 0) {
+            if (!connect(sock, t->ai_addr, t->ai_addrlen))
+                break;
+            close(sock);
+            sock = -1;
+        }
+    }
+
+    freeaddrinfo(res);
+    return sock;
+}
+
+UCC_CLASS_INIT_FUNC(ucc_tl_dpu_context_t,
+                    const ucc_base_context_params_t *params,
+                    const ucc_base_config_t *config)
+{
+    /* TODO: Need handshake with daemon for detection */
+    ucc_tl_dpu_context_config_t *tl_dpu_config =
+        ucc_derived_of(config, ucc_tl_dpu_context_config_t);
+
+    ucp_worker_params_t worker_params;
+    ucp_worker_attr_t   worker_attr;
+    ucp_params_t        ucp_params;
+    ucp_ep_params_t     ep_params;
+    ucp_ep_h            ucp_ep;
+    ucp_context_h       ucp_context;
+    ucp_worker_h        ucp_worker;
+    ucc_status_t        status = UCC_OK;
+    int ret;
+
+    /* Identify DPU */
+    char hname[256];
+    void *rem_worker_addr;
+    size_t rem_worker_addr_size;
+    int sockfd = 0, dpu_found = 0;
+
+    UCC_CLASS_CALL_SUPER_INIT(ucc_tl_context_t, tl_dpu_config->super.tl_lib,
+                              params->context);
+    memcpy(&self->cfg, tl_dpu_config, sizeof(*tl_dpu_config));
+
+    /* Find  DPU based on the host-dpu list */
+    gethostname(hname, sizeof(hname) - 1);
+    if (tl_dpu_config->use_dpu) {
```

Comment:
what the flag use_dpu does? enables/disables the whole TL? We don't need it in that case since TL selection is handled in CORE (UCC_TL_DPU_TUNE=0 - disable)